### PR TITLE
Enable thumbv7a-pc-windows-msvc target build end to end in rust/master

### PR DESCRIPTION
**Pull request by chandde**
With this PR, plus another commit https://github.com/rust-lang-nursery/compiler-builtins/commit/cf98161da7ed5217b6031796f0f60b4dd07148a4, I'm able to build the target thumbv7a-pc-windows-msvc successfully, and I'm able to use the stage2 artifacts to build arm32 projects. The commit in compiler_builtins is in release 0.1.14, the current cargo.lock in rust master still uses 0.1.12, so I bumped the compiler_builtins version in cargo.lock to 0.1.15

The command I used to build rust
```
c:\python27\python.exe x.py build --host x86_64-pc-windows-msvc --build x86_64-pc-windows-msvc --target thumbv7a-pc-windows-msvc --verbose
```

**Changes**
1. update cargolock to use compiler_builtins 0.1.15
2. handle libunwind in libtest for thumv7a the same as what we have for aarch64
3. in llvm codegen add a field in CodegenContext to carry the arch, so later in create_msvc_imps function, the arch can be used to check against "x86", instead of 32 pointer width. Apparently Thumv7a is handled differently than x86.

**Background**
I'm from Microsoft working on enabling Azure IoTEdge on ARM32 Windows IoTCore, Azure IoTEdge has a component called IoTEdged written in rust as a NT service running on Windows, so we need to enable rust on thumbv7a in order to have full IoTEdge. My colleague had made some heavy lifting and we've been using our private toolchain to build IoTEdged in our devops pipeline, because at that time we cannot build thumbv7a target end to end successfully. This change is a followup to enable the end to end build for thumbv7a-pc-windows-msvc target.

**Next step**
I'll submit more changes to have this target built nightly in rust/master, to achieve the same availability for aarch64-pc-windows-msvc, indexed here https://rust-lang.github.io/rustup-components-history/aarch64-pc-windows-msvc.html and can be manually installed. **Please do share what takes to make this happen, is there a formal process I need to follow\?**

### CODE REVIEW

**Comment by chandde** [comment]
Undo the changes to bump compiler_builtins to hopefully have a good build. The version bump perhaps is worth another separate PR, @alexcrichton ?

**Comment by mzji** [comment]
```
tidy check
[00:03:55] tidy error: /checkout/src/libtest/lib.rs:40: line longer than 100 chars
[00:04:00] some tidy checks failed
[00:04:00] 
```
@chandde I think you should wrap this line to not exceed 80 columns.

**Comment by mati865** [comment]
@chandde compiler-builtins are handled in https://github.com/rust-lang/rust/pull/60841 you should probably remove both update and revert commits.

**Comment by alexcrichton** [comment]
r=me with tidy passing, thanks @chandde!

File: src/libtest/lib.rs [hunk 1]
```diff
@@ -37,12 +37,12 @@ extern crate libc;
...
 // FIXME(#54291): rustc and/or LLVM don't yet support building with panic-unwind
-//                on aarch64-pc-windows-msvc, so we don't link libtest against
+//                on aarch64-pc-windows-msvc, or thumbv7a-pc-windows-msvc, so we don't link libtest against
 //                libunwind (for the time being), even though it means that
-//                libtest won't be fully functional on this platform.
+//                libtest won't be fully functional on these platforms.
 //
 // See also: https://github.com/rust-lang/rust/issues/54190#issuecomment-422904437
-#[cfg(not(all(windows, target_arch = "aarch64")))]
+#[cfg(not(any(all(windows, target_arch = "aarch64"), all(windows, target_arch = "arm"))))]
```
Comment (mati865) [inline src/libtest/lib.rs]: I think this would look simpler: `#[cfg(not(all(windows, any(target_arch = "aarch64", target_arch = "arm"))))]`.

**Comment by chandde** [comment]
thanks for the comments, I'm out attending some school activities with my son and won't have access to a pc until tonight. Will address all the comments later and get back.

**Comment by chandde** [comment]
@mati865 the update and revert have been forcefully removed in the commit history
@mzji @mati865 the line in libtest\lib.rs has been updated to `#[cfg(not(all(windows, any(target_arch = "aarch64", target_arch = "arm"))))]`, should have fixed the test failure and also look simpler

File: src/librustc_codegen_llvm/back/write.rs [hunk 2]
```diff
@@ -795,10 +795,10 @@ fn create_msvc_imps(
         return
     }
     // The x86 ABI seems to require that leading underscores are added to symbol
-    // names, so we need an extra underscore on 32-bit. There's also a leading
+    // names, so we need an extra underscore on x86. There's also a leading
     // '\x01' here which disables LLVM's symbol mangling (e.g., no extra
     // underscores added in front).
-    let prefix = if cgcx.target_pointer_width == "32" {
+    let prefix = if cgcx.target_arch == "x86" {
```
Comment (chandde) [inline src/librustc_codegen_llvm/back/write.rs]: FYI this change is solve this issue https://github.com/rust-lang/rust/issues/60866

**Comment by mzji** [comment]
> 
> 
> @mati865 the update and revert have been forcefully removed in the commit history
> @mzji @mati865 the line in libtest\lib.rs has been updated to `#[cfg(not(all(windows, any(target_arch = "aarch64", target_arch = "arm"))))]`, should have fixed the test failure and also look simpler

No, the tidy error line is the 40th line of `src/libtest/lib.rs`, looks like you didn't change it

File: src/libtest/lib.rs (hunk 1 above)
Comment (chandde) [reply src/libtest/lib.rs]: done!

**Comment by chandde** [comment]
> > @mati865 the update and revert have been forcefully removed in the commit history
> > @mzji @mati865 the line in libtest\lib.rs has been updated to `#[cfg(not(all(windows, any(target_arch = "aarch64", target_arch = "arm"))))]`, should have fixed the test failure and also look simpler
> 
> No, the tidy error line is the 40th line of `src/libtest/lib.rs`, looks like you didn't change it

Now it's done done

**Comment by chandde** [comment]
> r=me with tidy passing, thanks @chandde!

I don't understand the codes, did you say you're ok with the change once tidy passing? who's gonna sign off & merge? @alexcrichton

**Comment by jonas-schievink** [comment]
@chandde that's Mozilla "slang" that means your PR has been reviewed by whoever comments this and is ready to land when the remaining issues are addressed (in this case, the line length check).

Since that has been addressed:

@bors r=alexcrichton

**Comment by chandde** [comment]
thanks for your review! now I'd like to have master branch build the target thumbv7a-pc-windows-msvc nightly, so user can install the target using rustup, without locally building the tool chain, is there instruction how to do it? ideally the same availability for aarch64-pc-windows-msvc here https://rust-lang.github.io/rustup-components-history/aarch64-pc-windows-msvc.html

@alexcrichton @mzji @mati865 @jonas-schievink

**Comment by alexcrichton** [comment]
@chandde I think you'll basically want to grep around for aarch64-pc-windows-msvc in this repository and mirror its configuration, wherever aarch64 is mentioned thumbv7a probably also needs to be mentioned

**Comment by chandde** [comment]
> @chandde I think you'll basically want to grep around for aarch64-pc-windows-msvc in this repository and mirror its configuration, wherever aarch64 is mentioned thumbv7a probably also needs to be mentioned

I'll follow that, can you also point me to where the build actually happens for master, is it an azure devops pipeline, or... ? reading the build directly may also help.

**Comment by alexcrichton** [comment]
Currently the aarch64 target is built on AppVeyor, [configured here](https://github.com/rust-lang/rust/blob/d35181ad8785fa958e43580a29a982afe02c728f/appveyor.yml#L86-L95).

**Pull request merged**